ci/format.sh

```diff
@@ -1,100 +1,255 @@
+    if [[ $failed_java_checks -ne 0 ]]; then
+      error "$failed_java_checks Java files are formatted incorrectly."
+    fi
+  else
+    warning "Cannot find google-java-format, skipping Java file formatting!"
+  fi
+  return $failed_java_checks
+)
 
-set +e
-TRAILING_SPACES=$(git diff $DIFF_OPTS $BASE_SHA..HEAD -- $FILETYPES | xargs grep --line-number --with-filename '[[:blank:]]\+$')
-set -e
+# Strips off the "advice" at the end, since this script has different advice.
+function do_gn_check() {
+  local output
+  output="$("$SCRIPT_DIR/check_gn_format.py" --dry-run --root-directory . --gn-binary "third_party/gn/gn")"
+  local result=$?
+  echo "$output" | grep "ERROR"
+  return $result
+}
 
-if [[ ! -z "$TRAILING_SPACES" ]]; then
-  echo "$TRAILING_SPACES"
-  echo ""
-  echo "ERROR: Some files have trailing spaces. To fix, try something like \`find . -name "*.dart" -exec sed -i -e 's/\s\+$//' {} \;\`."
-  exit 1
-fi
+function check_gn_format() (
+  cd "$FLUTTER_DIR"
+  message "Checking GN formatting..."
+  if ! do_gn_check 1>&2; then
+    error "The gn file format check failed."
+    return 1
+  fi
+)
+
+function check_whitespace() (
+  local diff_opts=("-U0" "--no-color" "--name-only")
+  local filetypes="*.dart"
+  local trailing_spaces
+
+  message "Checking for trailing whitespace in $filetypes files..." 1>&2
+  set +e
+  trailing_spaces=$(git diff "${diff_opts[@]}" "$BASE_SHA" -- "$filetypes" | xargs grep --line-number --with-filename '[[:blank:]]\+$')
+  set -e
+
+  if [[ -n "$trailing_spaces" ]]; then
+    message "$trailing_spaces"
+    error "Whitespace check failed. The above files have trailing spaces."
+    return 1
+  fi
+)
+
+function fix_clang_format() {
+  local tmpfile
+  tmpfile=$(mktemp "fix_clang_format.XXXXXX")
+  if check_clang_format >"$tmpfile"; then
+    message "No C++/ObjC formatting issues found."
+  else
+    message "Fixing C++/ObjC formatting issues."
+    (
+      cd "$SRC_DIR/flutter"
+      patch -p0 <"$tmpfile"
+    )
+  fi
+  command rm -f "$tmpfile"
+}
+
+function fix_java_format() {
+  local tmpfile
+  tmpfile=$(mktemp "fix_java_format.XXXXXX")
+  if check_java_format >"$tmpfile"; then
+    message "No Java formatting issues found."
+  else
+    message "Fixing Java formatting issues."
+    (
+      cd "$SRC_DIR/flutter"
+      patch -p0 <"$tmpfile"
+    )
+  fi
+  command rm -f "$tmpfile"
+}
 
-# Check GN format consistency
-./ci/check_gn_format.py --dry-run --root-directory . --gn-binary "third_party/gn/gn"
+function fix_gn_format() (
+  cd "$FLUTTER_DIR"
+  message "Fixing GN formatting..."
+  # Check GN format consistency and fix it.
+  if ! "$SCRIPT_DIR/check_gn_format.py" --root-directory . --gn-binary "third_party/gn/gn"; then
+    error "The GN file format fix failed."
+    return 1
+  fi
+)
+
+function fix_whitespace() {
+  if ! check_whitespace; then
+    message "Fixing trailing whitespace problems."
+    find "$FLUTTER_DIR" -type f -name "*.dart" -print0 | xargs -0 sed -i -e 's/\s\+$//'
```

Comment:
Can we make this find only files that have changes in this patch?  I generally don't want to fix files that I'm not touching anyway. This will also be relatively slow or some people, especially as web continues to add files.

Reply:
Yes, I could do that.

TBH, if you want, on Friday I fooled around with rewriting the formatting script in Dart.  It uses the `process_runner` package to shard out the clang and Java formatting, and runs about 3.5 times faster (4s vs 14s). I didn't include it here because I decided that it was more complicated, and thus more complicated to maintain, so I stuck with Bash, but Bash is not the most ideal programming language...

Here's a diff with the implementation: https://github.com/flutter/engine/compare/master...gspencergoog:new_format_script?expand=1

Reply:
I think if we rewrote it in dart that would be preferable. We already have some dart based tools for this kind of thing, and bash gets tricky in ways that I don't think too many people on the team are aware of (myself included).

Reply:
That said, for now this is probably fine - the slow down would only be or people running --fix locally. I you want to land this as is LGTM, but it'd be nice to push forward with the Dart refactor.